### PR TITLE
feature/employee - deleteEmployee

### PR DESCRIPTION
**Pull request by dennisvoelkl**

### CODE REVIEW

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/Employee.java [hunk 1]
```diff
@@ -3,6 +3,4 @@
 public record Employee(
         String id,
         String name
-) {
-    
-}
+) {}
```
Comment (ToLa09) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/Employee.java:6]: Missing linebreak at the end.

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeRepo.java [hunk 2]
```diff
@@ -16,7 +17,13 @@ public Employee addEmployee(Employee saveEmployee) {
...
 
+    public String deleteEmployee(String id) {
+        for (Employee employeeToDelete : employeeList)
+            if (Objects.equals(employeeToDelete.id(), id)) {
+                employeeList.remove(employeeToDelete);
+                return employeeToDelete.name() + " erfolgreich gelöscht";
+            }
+        throw new NoSuchElementException("Employee konnte nicht gelöscht werden");
+    }
+}
```
Comment (ToLa09) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeRepo.java:29]: Missing linebreak at the end.

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java [hunk 3]
```diff
@@ -25,4 +25,8 @@ public Employee addEmployee(NewEmployee newEmployee) {
         Employee saveEmployee = newEmployee.withId(id);
         return employeeRepo.addEmployee(saveEmployee);
     }
+
+    public String deleteEmployee(String id) {
+        return employeeRepo.deleteEmployee(id);
+    }
 }
```
Comment (ToLa09) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java:32]: Missing linebreak at the end.

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/NewEmployee.java [hunk 4]
```diff
@@ -6,4 +6,4 @@ public record NewEmployee(
     public Employee withId(String id) {
         return new Employee(id, name);
     }
-}
+}
```
Comment (ToLa09) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/NewEmployee.java:9]: Missing linebreak at the end.

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java [hunk 5]
```diff
@@ -25,4 +25,8 @@ public Employee addEmployee(NewEmployee newEmployee) {
         Employee saveEmployee = newEmployee.withId(id);
```
Comment (ToLa09) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java:25]: maybe write `Employee saveEmployee = new Employee(id, newEmployee.name)`, so you don't need the additional method in the NewEmployee Record.

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java [hunk 6]
```diff
@@ -62,4 +61,37 @@ void addEmployeeReturnsEmployeeWithID() throws Exception {
...
+        //given
+        String content = mockMvc.perform(MockMvcRequestBuilders.post("/api/employees")
+
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content("""
+                                        {
+                                            "name": "Hasi"
+                                        }
+                                        """))
+                .andExpect(status().isOk())
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java:76]: POST-Request should return Status 201 (isOk = Status 200)

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java [hunk 7]
```diff
@@ -62,4 +61,37 @@ void addEmployeeReturnsEmployeeWithID() throws Exception {
...
+                .andReturn().getResponse().getContentAsString();
+        System.out.println("Content: " + content);
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java:78]: system out prints shouldn't be in the final code

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java [hunk 8]
```diff
@@ -62,4 +61,37 @@ void addEmployeeReturnsEmployeeWithID() throws Exception {
...
+        Employee employee = objectMapper.readValue(content, Employee.class);
+        System.out.println("employee: " + employee);
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java:80]: system out prints shouldn't be in the final code

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java [hunk 9]
```diff
@@ -62,4 +61,37 @@ void addEmployeeReturnsEmployeeWithID() throws Exception {
...
+        //when
+        mockMvc.perform(MockMvcRequestBuilders.get("/api/employees"))
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java:82]: get request isn't needed in this Test

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java [hunk 10]
```diff
@@ -62,4 +61,37 @@ void addEmployeeReturnsEmployeeWithID() throws Exception {
...
+                        "name": "Hasi"
+                    }]
+                    """.replace("<id>", employee.id())));
+
+        //then
+        mockMvc.perform(MockMvcRequestBuilders.delete("/api/employees/" + employee.id()))
+                .andExpect(status().isOk())
+                .andExpect(content().string("Hasi erfolgreich gelöscht"));
+    }
 }
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeControllerTest.java:97]: Missing linebreak at the end.

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeRepoTest.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    @DirtiesContext
+    void deleteEmployeeThrowNoSuchElementException() {
+        Exception exception = assertThrows(NoSuchElementException.class, () -> {
+            employeeRepo.deleteEmployee(null);
+        });
+        String expectedMessage = "Employee konnte nicht gelöscht werden";
+        String actualMessage = exception.getMessage();
+        assertTrue(actualMessage.contains(expectedMessage));
+    }
+}
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeRepoTest.java:21]: Missing linebreak at the end.

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java [hunk 12]
```diff
@@ -39,4 +38,16 @@ void addEmployeeTEST() {
...
+    void deleteEmployeeTEST() {
+        //given
+        String testToDeleteString = "UUIDFromController";
+        //when
+        String actual = employeeService.deleteEmployee(testToDeleteString);
+        //then
+        String expected = null;
+        assertEquals(expected, actual);
+    }
 }
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java:53]: Missing linebreak at the end.

File: frontend/src/employee/EmployeeCard.tsx [hunk 13]
```diff
@@ -3,10 +3,15 @@ import {EmployeeModel} from './EmployeeModel'
...
+
 export default function EmployeeCard(props: employeeProps) {
+
     return <>
-        <li>Name: {props.employee.name} ID: {props.employee.id}</li>
+        <li>Name: {props.employee.name} ID: {props.employee.id}
+            <button onClick={() => props.deleteEmployee(props.employee.id)}>Delete Employee</button>
+        </li>
     </>
 }
```
Comment (ToLa09) [inline frontend/src/employee/EmployeeCard.tsx:17]: Missing linebreak at the end.

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java [hunk 14]
```diff
@@ -39,4 +38,16 @@ void addEmployeeTEST() {
...
+
+    @Test
+    @DirtiesContext
+    void deleteEmployeeTEST() {
+        //given
+        String testToDeleteString = "UUIDFromController";
+        //when
+        String actual = employeeService.deleteEmployee(testToDeleteString);
+        //then
+        String expected = null;
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java:50]: expected String should be `employeeToDelete.name() + " erfolgreich gelöscht"`

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java [hunk 15]
```diff
@@ -39,4 +38,16 @@ void addEmployeeTEST() {
...
         assertEquals(employeeWithId, actual);
     }
+
+    @Test
+    @DirtiesContext
+    void deleteEmployeeTEST() {
+        //given
+        String testToDeleteString = "UUIDFromController";
+        //when
+        String actual = employeeService.deleteEmployee(testToDeleteString);
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java:48]: mocking is missing, thats why employee.deleteEmployee returns null

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java [hunk 16]
```diff
@@ -39,4 +38,16 @@ void addEmployeeTEST() {
         //then
         assertEquals(employeeWithId, actual);
     }
+
+    @Test
+    @DirtiesContext
```
Comment (ToLa09) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java:43]: DirtiesContext is used for integration tests, can be deleted

File: frontend/src/employee/EmployeeOverview.tsx [hunk 17]
```diff
@@ -7,25 +7,32 @@ export default function EmployeeOverview() {
...
     const getAllEmployees = () =>
-        axios.get('/api/employees')
+        axios.get(baseUrl)
             .then((answer) => {
                 setEmployees((answer.data))
             })
             .catch((error) => console.log(error))
 
+    function deleteEmployee(id: string) {
+        console.log("id: " + id)
```
Comment (ToLa09) [inline frontend/src/employee/EmployeeOverview.tsx:24]: there should be no console.logs in production code

**Review by ToLa09 — CHANGES_REQUESTED** [review]

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeRepo.java (hunk 2 above)
Comment (chris-yooo) [reply backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeRepo.java:29]: ?

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java (hunk 16 above)
Comment (chris-yooo) [reply backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeServiceTest.java:43]: ok removed

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java (hunk 3 above)
Comment (chris-yooo) [reply backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java:32]: ?

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java (hunk 5 above)
Comment (chris-yooo) [reply backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/employee/EmployeeService.java:25]: we have extra outsourced this here like daniels annotation

**Review by Kreyno93 — APPROVED** [review]

**Pull request merged**